### ISSUE TITLE
Publish past experiment directly from Taskcluster

### ISSUE THREAD

**Issue by La0** [mozilla/firefox-translations-training#312]
Depends #309 

We need a new publication mode that relies only on a Taskcluster task group ID to find all relevant tasks and their artifacts (configurations, logs and metrics) and publish them on Weight & Biases.

This will not run inside Taskcluster, but help to publish existing experiments into Weight & Biases without relying on model/experiment/logs directory.

Past experiments to publish:
- [en-hu](https://firefox-ci-tc.services.mozilla.com/tasks/groups/aY25-4fXTcuJNuMcWXUYtQ)
- [lt-en 1](https://firefox-ci-tc.services.mozilla.com/tasks/groups/EW7qV3U5SBSjegnTxGZHkw)
- [lt-en 2](https://firefox-ci-tc.services.mozilla.com/tasks/groups/V-OmRM1yS_GwESTVDPl0VQ)
- [lt-en 3](https://firefox-ci-tc.services.mozilla.com/tasks/groups/K1iHndFUSxSEDRLg_H9l1A)
- [lt-en 4](https://firefox-ci-tc.services.mozilla.com/tasks/groups/DXbS0zreSGSVYloAF8gwJg)

**Comment by eu9ene** [comment]
We should also make sure that we read the experiment name from the train action config in the same way we do with the config on disk. 

Since most experiments required changing code and retrying things we'll need to traverse dependent Taskcluster tasks starting from the final "all" task.

**Comment by eu9ene** [comment]
You will likely want to re-purpose this script that already does training logs and evals downloading https://github.com/mozilla/firefox-translations-training/pull/335/files

**Comment by eu9ene** [comment]
I've been thinking how to treat experiments with the same name that span across multiple task groups. Usually the scenario is to create a git branch, do some fixes, run Train action from a push task group, look at the results, do more fixes or/and rerun with a different experiment config (see [docs](https://github.com/mozilla/firefox-translations-training/blob/main/docs/task-cluster.md)). Taskcluster doesn't enforce changing experiment name in the config. From developer's perspective all this can be a part of a larger experiment but technically each Train action and a task group is a different experiment because it runs with different code and experiment config. So I'm thinking we'll need to add a suffix with task group id to the experiment name (run group) in W&B:

For example for https://wandb.ai/teklia/en-ru experiment `ny-retraining` if run on Taskcluster would become:

`ny-retraining_<task-group-id>`

Example TC experiment:
https://firefox-ci-tc.services.mozilla.com/tasks/groups/Tkrf0fGBQEO6kH-gSKp5lg

would be named:

`opustrainer_no_aug_Tkrf0fGBQEO6kH-gSKp5lg`

where `opustrainer_no_aug` corresponds to `experiment.name` from the train action config https://firefox-ci-tc.services.mozilla.com/tasks/Tkrf0fGBQEO6kH-gSKp5lg
and `Tkrf0fGBQEO6kH-gSKp5lg` is a task group id for the parent task group of the pipeline steps https://firefox-ci-tc.services.mozilla.com/tasks/groups/Tkrf0fGBQEO6kH-gSKp5lg

**Comment by La0** [comment]
Here is a summary of the steps to implement using unauthenticated Taskcluster Python client that will:
1. load training logs and metrics from a taskcluster group, pretty similar to the existing [taskcluster_downloader.py](https://github.com/mozilla/firefox-translations-training/blob/main/utils/taskcluster_downloader.py) script
2. parse them using @vrigal implementation of `TrainingParser` in this repo
3. publish under a project & experiment name produced from the parent task parameters

The only secret needed is a W&B token as all Taskcluster data is openly available

## 0: pre-flight

- init Taskcluster queue client

```python
import taskcluster

queue = taskcluster.Queue({"rootUrl": 'https://firefox-ci-tc.services.mozilla.com'})
```


## 1: check group 

- Query `queue.getTaskGroup` API call
    - abort if it raises 404
- List all **completed** (from `status.state`) tasks with `queue.listTaskGroup` API call
    - abort if none found


## 2: get project & experiment name

- use task details using group id
    - this is the task trigerred by a human operator
    - it holds all the initial parameters

```python
parent = queue.task(group_id)
config = parent["extra"]["action"]["context"]["input"]
experiment = config["experiment"]
project = f"{experiment['src']}-{experiment['trg']}"
name = f"{experiment['name']}_{group_id}"
```

- `project` is source + target languages
- experiment `name` is the one from config + group ID to get a unique name

## 3: load training logs

- Find all tasks in group from step 1
    - whose `task.tags.kind` starts with `train` or `finetune`
    - and do not have `vocab` in `task.tags.kind`
- Abort if none found
- Download and parse every listed task logs
    - using [downloadArtifactToBuf](https://github.com/taskcluster/taskcluster/blob/main/clients/client-py/taskcluster/download.py#L51)
    - artifact name is `public/build/train.log`

```python
from taskcluster.download import downloadArtifactToBuf

training_tasks = [
    task_status["status"]["taskId"]
    for task_status in group["tasks"]
    if task_status["task"]["tags"]["kind"].startswith('train')
]

for task_id in training_tasks:
    log, _ = downloadArtifactToBuf(
        taskId=task_id,
        name="public/build/train.log",
        queueService=queue,
    )

    print(log.tobytes().decode())
```

## 4: load metrics

- Find all tasks in group from step 1
    - whose `task.tags.kind` starts with `evaluate`
- list all artifacts on every task that  whose name ends in `.metrics`
- Download and parse every metrics log (similar to step 3)


```python
evaluate_tasks = [
    task_status["status"]["taskId"]
    for task_status in group["tasks"]
    if task_status["task"]["tags"]["kind"].startswith('evaluate')
]
artifacts = [
    (task_id, a["name"])
    for task_id in evaluate_tasks
    for a in queue.listLatestArtifacts(task_id)["artifacts"]
    if a["name"].endswith(".metrics")
]
```



## 5: parse & publish

Using [TrainingParser](https://github.com/mozilla/firefox-translations-training/blob/main/tracking/translations_parser/parser.py) to load all training metrics from log files & chrf files

Use project & experiment name from step 2 to publish in the correct W&B project, under a new experiment.

Finally display a link towards the W&B experiment

**Comment by eu9ene** [comment]
You'll also need to traverse the tasks in the task group to find cached steps from other task groups and publish those task groups too.

> Since most experiments required changing code and retrying things we'll need to traverse dependent Taskcluster tasks starting from the final "all" task.

**Comment by La0** [comment]
I don't know what is a cached step in this context: We'll discuss later today

**Comment by La0** [comment]
Thanks to our discussion on Monday, I was able to build the following POC script that recursively browse the dependencies of initial training tasks found in a group, and find other training tasks in these groups.

The dependencies are available through the `queue.task` TC Python API call, then the group ID is retrieved using `queue.status`.


```python
import taskcluster

queue = taskcluster.Queue({"rootUrl": 'https://firefox-ci-tc.services.mozilla.com'})


def find_training_tasks(group_id):
    return [
        ts
        for ts in queue.listTaskGroup(group_id)["tasks"]
        if ts["status"]["state"] == "completed" and  ts["task"]["tags"]["kind"].startswith('train')
    ]

def list_parent_training_tasks(task_id, known=[]):
    task = queue.task(task_id)

    # Browse task parents
    for parent_id in task['dependencies']:
        parent_status = queue.status(parent_id)
        gid = parent_status["status"]["taskGroupId"]
        if gid in known:
            continue

        # Provide training tasks from new parent group
        yield from find_training_tasks(gid)
        known.append(gid)

        # Recurse
        yield from list_parent_training_tasks(parent_id, known)

def main(group_id):

    training = find_training_tasks(group_id)
    for ts in training:
        print("Top level training", ts["status"]["taskId"], ts["task"]["tags"]["label"])

        for ts_ in list_parent_training_tasks(ts["status"]["taskId"]):
            print("SUB level training", ts_["status"]["taskId"], ts_["task"]["tags"]["label"])
 
```

By calling `main('aY25-4fXTcuJNuMcWXUYtQ')`, we get the following output that display the initial training task in the top group, then all the training tasks discovered while browsing the parent tasks and their groups.

```
Top level training VyiNNDG3TqSLM00TODS_6w train-student-en-hu
SUB level training LZhdArX0SE2y51JGVM3XlQ train-teacher-en-hu-1/2
SUB level training KZPjvTEiSmO--BXYpQCNPQ train-teacher-en-hu-2/2
SUB level training McHmCq2cSVmR9OZXJrC4cw train-backwards-en-hu
SUB level training Vof1ufPxTOyG3OiSj4oHwQ train-vocab-en-hu
SUB level training bpUaubrpQgOPegv8ZH9xTg train-teacher-en-hu-2/2
SUB level training YQUp-mUgTNSsJS6RjVSUrQ train-teacher-en-hu-1/2
```

These steps should be run first to get a list of group ID and their training tasks, then apply steps 1 to 4 for each training task so they get published on W&B

**Comment by vrigal** [comment]
@eu9ene I almost finished the first part (handling a single group) and would like to discuss 2 things:
* Sometimes evaluation seems to be splitted into 2 tasks (e.g. `evaluate-student-sacrebleu-aug-lt-en-1/2` and `evaluate-student-sacrebleu-aug-lt-en-2/2`). Which one should be published ?
* Training tasks names do not match evaluation task names (quantize -> quantized, finetune -> finetuned). Should I add a rule to support the `ed` prefix ?

**Comment by eu9ene** [comment]
> @eu9ene I almost finished the first part (handling a single group) and would like to discuss 2 things:
> 
>     * Sometimes evaluation seems to be splitted into 2 tasks (e.g. `evaluate-student-sacrebleu-aug-lt-en-1/2` and `evaluate-student-sacrebleu-aug-lt-en-2/2`). Which one should be published ?
> 
>     * ~Training tasks names do not match evaluation task names (quantize -> quantized, finetune -> finetuned). Should I add a rule to support the `ed` prefix ?~
>       edit: I discussed with Bastien and this is doable by loading dependencies among tasks, which will be cleaner !

The 1/2 2/2 evaluation happens only for the teachers because we train two of them. Those metrics will correspond to the teacher models that you see in dependencies (for examples ["evaluate-teacher-flores-flores_aug-typos_devtest-lt-en-1/2"](https://firefox-ci-tc.services.mozilla.com/tasks/Jdd7r3_ATveOiayB9ZnpUA) evaluates ["train-teacher-lt-en-1/2"](https://firefox-ci-tc.services.mozilla.com/tasks/e8S8_5xhQeaNpSaSMej9CA)). All other models are trained once. We might also experiment with training 4 teachers in the future.

**Comment by vrigal** [comment]
@eu9ene I made the `publish_group_logs` function generic, so the behavior is the same than for experiments.
This is not ideal for the Taskcluster approach, as it requires to build a structure with logs of the evaluation tasks (we could direcly read `.metrics` files here).
It is functional like this, we can open an issue once we merge this IMO to improve this specific part of the code.

**Comment by eu9ene** [comment]
Here is the current list of top-level groups to publish. We should do this when the issues with consistency are addressed (#452 #448 #449 ). You can also check this [doc](https://docs.google.com/spreadsheets/d/1iyKKWVYXV68sAKwzU9Y6f6KNRnvQQ7HJ7431TSfr5BU/edit#gid=0) GroupId column for recent updates.

```
KsSrCXPtRzCie4wkejInsA
KRXRJ_lTSWWrv0F20lORCQ
ODv_K3kWTBuFdGdY3-WW2A
BgvHqoMIRRGUJct7IkokQQ
Wz8KJcqPTAOFhV-YaRGJ0g
Cjzd04QgRDaL-w0AAJ0TlQ
YxSGB_9sTTy7d6HiWHxtEA
EFSRZZ-MSzSA5FL49Ja_FQ
FgNRsJFdQ1CkjT8I20utGg
daBcFxZfS_WDp_31N5zrDw
aY25-4fXTcuJNuMcWXUYtQ
EW7qV3U5SBSjegnTxGZHkw
DXbS0zreSGSVYloAF8gwJg
V-OmRM1yS_GwESTVDPl0VQ
K1iHndFUSxSEDRLg_H9l1A
Tkrf0fGBQEO6kH-gSKp5lg
duPWN5eVQ3uFV7GKdyTVnQ
ZZStzQpXTJCZ-RGjbU3yHA
eRqJKDrzRTaj2dhpSHfx2A
ZaKoAaeqR1GLEnMWRb2tDQ
H396IbuxSAWjhjpBvg_azw
```

**Comment by eu9ene** [comment]
Ok, the experiments are now published but there are some issues I found: #518
